### PR TITLE
CE Metadata changes 

### PR DESCRIPTION
**Pull request by fearthecowboy**
Changes:
 - metadata changes that were in the spec document
 - validation for object model added and enabled (now you will see errors and warnings when artifacts files are not conformant)
 - activation was refactored to make it easier to generate more things in activation (msbuild, aliases)
 - custom post-install action for espressif-idf artifact
 - much better messages
 - some misc cleanups in the code 
 
Registry to go along with the changes: 
  https://github.com/microsoft/vcpkg-ce-catalog/tree/metadata-changes

### CODE REVIEW

File: ce/ce/amf/demands.ts [hunk 1]
```diff
@@ -298,33 +290,27 @@ export class Unless extends DemandBlock implements AlternativeFulfillment {
...
-            filtered = filter(this.select || '', result.log);
-            filtered.$0 = item;
-
-            // if we have a match expression, let's check it.
-            if (this.matches && !safeEval(this.matches, filtered)) {
-              continue; // not a match, move on
-            }
-
-            // it did match, or it's just presence check
+            this.discoveredData['$0'] = item.toString();
```
Comment (strega-nil-ms) [inline ce/ce/amf/demands.ts:302]: I don't understand why this line is important given 293; isn't this just setting `this.discoveredData['$0']` to exactly what it already is?

File: ce/ce/amf/demands.ts [hunk 2]
```diff
@@ -298,33 +290,27 @@ export class Unless extends DemandBlock implements AlternativeFulfillment {
...
-          this.setData(filtered);
-          if (this.run) {
+          this.discoveredData = { $0: item.toString() };
+          const run = this.run?.replace('$0', item.toString());
 
-            const commandline = cmdlineToArray(this.run.replace('$0', item.toString()));
-            const result = await execute(resolve(commandline[0]), commandline.slice(1));
-            if (result.code !== 0) {
+          if (run) {
+            const result = await this.runCmd(run);
```
Comment (strega-nil-ms) [inline ce/ce/amf/demands.ts:297]: why are we running binaries? what part of the spec does this implement?

File: ce/ce/util/safeEval.ts [hunk 3]
```diff
@@ -25,3 +25,116 @@ export function createSandbox(): <T>(code: string, context?: any) => T {
     return sandbox[response];
   };
 }
+
+export const safeEval = createSandbox();
+
+function isValue(cursor: Scanner) {
```
Comment (strega-nil-ms) [inline ce/ce/util/safeEval.ts:31]: where are these functions called? How do we know what's being passed to `safeEval`?

File: ce/ce/cli/markdown-table.ts [hunk 4]
```diff
@@ -2,6 +2,36 @@
...
+      chars: {
+        'top': '', 'top-mid': '', 'top-left': '', 'top-right': ''
+        , 'bottom': '', 'bottom-mid': '', 'bottom-left': '', 'bottom-right': ''
+        , 'left': '', 'left-mid': '', 'mid': '', 'mid-mid': ''
+        , 'right': '', 'right-mid': '', 'middle': ''
+      }
+    }
+  }),
+  gfm: true,
+});
```
Comment (strega-nil-ms) [inline ce/ce/cli/markdown-table.ts:34]: what's up with all these options?

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]
I can't tell what is going on with this PR; it seems really massive and it touches a lot of different features. Would it be possible to split it into multiple PRs?

File: ce/ce/yaml/EntitySequence.ts [hunk 5]
```diff
@@ -17,6 +17,7 @@ export /** @internal */ class EntitySequence<TElement extends Yaml<YAMLDictionar
     return new YAMLDictionary();
   }
 
+
```
Comment (ras0219-msft) [inline ce/ce/yaml/EntitySequence.ts:20]:
```suggestion

```

File: ce/ce/util/linq.ts [hunk 6]
```diff
@@ -122,11 +114,11 @@ export function keys<K, T, TSrc>(source: any): Iterable<any> {
 
 /** returns an IterableWithLinq<> for keys in the collection */
 function _keys<K, T>(source: Map<K, T> | null | undefined): IterableWithLinq<K>
-function _keys<T, TSrc extends Dictionary<T>>(source: Dictionary<T> | null | undefined): IterableWithLinq<string>
+function _keys<T, TSrc extends Record<string, T>>(source: Record<string, T> | null | undefined): IterableWithLinq<string>
 function _keys<T, TSrc extends Array<T>>(source: Array<T> | null | undefined): IterableWithLinq<number>
 function _keys<K, T, TSrc>(source: any | undefined | null): IterableWithLinq<any>
 function _keys<K, T, TSrc>(source: any): IterableWithLinq<any> {
-  //export function keys<K, T, TSrc extends (Array<T> | Dictionary<T> | Map<K, T>)>(source: TSrc & (Array<T> | Dictionary<T> | Map<K, T>) | null | undefined): IterableWithLinq<IndexOf<TSrc>> {
+  //export function keys<K, T, TSrc extends (Array<T> | Record<string,T> | Map<K, T>)>(source: TSrc & (Array<T> | Record<string,T> | Map<K, T>) | null | undefined): IterableWithLinq<IndexOf<TSrc>> {
```
Comment (ras0219-msft) [inline ce/ce/util/linq.ts:121]:
```suggestion
```

File: ce/ce/amf/demands.ts (hunk 2 above)
Comment (dan-shaw) [reply ce/ce/amf/demands.ts:297]: Demands can run commands, for example to get a version string of a tool.

File: ce/ce/util/safeEval.ts (hunk 3 above)
Comment (dan-shaw) [reply ce/ce/util/safeEval.ts:31]: `safeEval` evaluates `is` statements. `isValue` helps determine if these statements are valid expressions.

File: ce/ce/amf/demands.ts (hunk 2 above)
Comment (BillyONeal) [reply ce/ce/amf/demands.ts:297]: What?! Is there a spec for such a thing?

File: ce/ce/cli/markdown-table.ts (hunk 4 above)
Comment (fearthecowboy) [reply ce/ce/cli/markdown-table.ts:34]: It's using a markdown renderer to generate tables. This sets the renderer options to create pretty tables.

File: ce/ce/artifacts/activation.ts [hunk 7]
```diff
@@ -1,67 +1,561 @@
...
 import { Session } from '../session';
-import { linq } from '../util/linq';
+import { isIterable } from '../util/checks';
+import { linq, Record } from '../util/linq';
+import { Queue } from '../util/promise';
 import { Uri } from '../util/uri';
 import { toXml } from '../util/xml';
 import { Artifact } from './artifact';
 
+function findCaseInsensitiveOnWindows<V>(map: Map<string, V>, key: string): V | undefined {
```
Comment (BillyONeal) [inline ce/ce/artifacts/activation.ts:20]: Ouch this really needs our almost_canonical. No change requested.

https://github.com/microsoft/vcpkg-tool/blob/3da30bc77bda131562fd032ce5c4334c9f398cef/src/vcpkg/base/files.cpp#L3087-L3100

File: ce/ce/artifacts/activation.ts [hunk 8]
```diff
@@ -77,90 +571,197 @@ export class Activation {
...
+      undo[variable] = originalEnvironment[variable] || '';
+    }
 
-  /** a collection of arbitrary properties from artifacts. useful for msbuild */
-  properties = new Map<string, Array<string>>();
+    // .defines get compiled into a single environment variable.
+    if (this.definesCount) {
+      let defines = '';
+      for await (const [name, value] of this.defines) {
+        defines += value ? `-D ${name}=${value} ` : `-D ${name} `;
```
Comment (BillyONeal) [inline ce/ce/artifacts/activation.ts:613]:
```suggestion
        defines += value ? `-D${name}=${value} ` : `-D${name} `;
```

Traditionally there isn't supposed to be a space here although I suspect most compilers don't care.

**Review by BillyONeal — APPROVED** [review]

**Review by strega-nil-ms — APPROVED** [review]
I don't see anything obvious where I go "this is not good", so I'm a soft approve. I still don't know TS to be comfortable saying 'yeah this looks great'

File: ce/ce/artifacts/activation.ts (hunk 8 above)
Comment (fearthecowboy) [reply ce/ce/artifacts/activation.ts:613]: I will defer to you on that - I don't know for certain.

**Pull request merged**